### ISSUE TITLE
Switch to CMake

### ISSUE THREAD

**Issue by tstenner** [sccn/labstreaminglayer#207]
As previously discussed in #203 I think it's a good idea to switch to CMake for liblsl and the apps. With modern CMake (e.g. [1](https://rix0r.nl/blog/2015/08/13/cmake-guide/), [2](https://schneide.wordpress.com/2016/04/08/modern-cmake-with-target_link_libraries/)) it's possible to keep the current tree with subfolders for apps and the library and have one build for everything.
Currently, I've created CMakeLists.txt for [the root directory](https://github.com/tstenner/labstreaminglayer/blob/cmake_wip/CMakeLists.txt), [all apps](https://github.com/tstenner/labstreaminglayer/blob/cmake_wip/Apps/CMakeLists.txt), the [library](https://github.com/tstenner/labstreaminglayer/blob/cmake_wip/LSL/liblsl/CMakeLists.txt) and some apps.

Because all targets are available to CMake, linking to liblsl is done with a simple `target_link_libraries(targetname PUBLIC LSL::lsl)`, CMake takes care of include directories, linker flags, defines and additional compiler options. This information is also exported, so that external projects can re-import the library with everything necessary ([example for a CMakeLists.txt importing LSL](https://github.com/tstenner/labstreaminglayer/blob/cmake_wip/OutOfTreeTest/CMakeLists.txt)).
Still, building liblsl (including the bundled Boost) takes less than 2 minutes on appveyor or a 6 year old laptop, so downloading a pre-compiled library doesn't save much time but could cause problems with different compiler versions / compile flags.

For distribution the CMake configuration defines an installation target that might create a directory tree like this:
<pre>LSL64_Release_MSVC14
├── AppX
│   ├── AppX.exe
│   ├── AppX.bat
│   └── AppX_configuration.ini
├── AppY
│   ├── AppY.exe
│   ├── AppY.bat
│   ├── AppY_conf.exe
│   └── example.png
├── examples
│   ├── CppReceive.exe
│   ├── CppReceive.bat
│   ├── CppSendRand.exe
│   ├── CppSendRand.bat
│   ├── SendDataC.exe
│   └── SendDataC.bat
└── LSL
    ├── cmake
    │   ├── LSLAppBoilerplate.cmake
    │   ├── LSLConfig.cmake
    │   └── [snip]
    ├── include
    │   ├── lsl_c.h
    │   └── lsl_cpp.h
    └── lib
        ├── liblsl64_MSVC14.dll
        ├── liblsl64_MSVC14.lib
        ├── lslboost.lib
        ├── Qt5Xml.dll
        ├── Qt5Gui.dll
        ├── [more Qt]
        ├── libboost-system-msvc14-mt.dll
        └── [more boost]</pre>

On Unix systems (Linux+OS X) the library path is changed to include `../LSL/lib/` and `./lib/` so common libraries (Qt, Boost) can be distributed in a single library directory wheras specific libraries can be put in a subdirectory of the App dir. On Windows, a `.bat`-file is autogenerated that adds these library paths to the `PATH` variable and starts the program.

For Qt I haven't tested this yet, but adding the required libraries to `LSL/lib` and adding a [`Qt.conf`](http://doc.qt.io/qt-5/qt-conf.html) that tells Qt where to find the platform plugins.

For distribution, the subfolders have to be zipped up, yielding a directory (`LSL64_Release_MSVC14` / `LSL_Release_Win64`?) with one zip file per App and one for the shared libraries.

**Comment by dmedine** [comment]
This looks fantastic! Excellent work. Please submit a pull request and I will evaluate everything. This may take some time as there are many things to test and I may want to change a few details.

I would also point out that a few days ago the boost community announced that they will be moving to CMake by the end of the year. I think that alone is indication that it is time to do this.

**Comment by kblomqvist** [comment]
Could you provide a link to that announcement by boost community?

**Comment by tstenner** [comment]
There are still some things I'm working on (adding the MSVC version to DLLs, putting Boost/Qt libraries in LSL/lib, automatically running `windeploy`/`macdeployqt` for binaries after installing) and I'll have to separate the BrainAmp changes from the CMake changes, but I hope to get it done today or tomorrow.

With Boost I think it'll be some time before we can use the CMake builds as at least the Linux build environment on Travis CI is a bit older. This shouldn't really matter, as (starting with CMake 3.5) we can import use Boost like this:

    FindPackage(Boost REQUIRED COMPONENTS thread iostreams)
    target_link_libraries(target PUBLIC Boost::thread Boost::iostreams

**Comment by dmedine** [comment]
@kblomqvist I heard it on the boost-users mailing list. It was from 2 days ago (July 18) so it hasn't been archived yet, but in two weeks or so you can search for it here https://lists.boost.org/boost-users/.

@tstenner I referred to this switch with boost just to illustrate the viability of CMake. I don't expect any linking issues apart from the normal headaches.

**Comment by cboulay** [comment]
@tstenner ,

Thanks very much for doing this. It's something I've wanted to do for a while but never found time. 

Some comments:

1. I would prefer it if the default was to search for system-installed third party libraries then, if not found, download the necessary libraries.
    * The goal here is to save compile time, hard drive space, and confusion for people like me who develop for multiple different projects which all happen to have the same dependencies.
    * We should use official Find<lib>.cmake when available, otherwise write our own, possibly borrowing from other similar files found on the net.
        * When writing our own, we can search in the default install directories for libraries that have binary installers on Windows. (e.g., [boost](https://sourceforge.net/projects/boost/files/boost-binaries/), 
    * After attempting to find a package using official .cmake files, we can do `IF(NOT <lib>-FOUND)` and then use [`ExternalProject_Add`](https://cmake.org/cmake/help/latest/module/ExternalProject.html). I have an example of how to use that for OpenCV [here](https://github.com/cboulay/PSMoveService/blob/master/cmake/ThirdParty.cmake#L54-L127).
2. For apps, only LabRecorder should be built by default. All the others should be opt-in, or maybe auto-opt-in if the necessary dlls are found already in the expected location.
    * Expected location should be overridable with something like -DDEV_SDK_DIR=C:\Manufacturer\SDK
    * If opting in but dlls aren't found, then they can be automatically downloaded
3. Almost every app that currently uses boost can probably replace the boost dependency with STL and Qt. Normally I'd say boost is a better dependency to have than Qt, except that Qt is probably being used anyway for just about every app. I'm not suggesting you go through all the apps and rewrite them to use Qt instead of boost. I'm just pointing out that, for many users, installing boost could be completely unnecessary especially if LSL continues to depend on its own custom boost.
4. I prefer to make the apps as portable as possible.
    * On Mac, this requires some careful RPATH work and placing the dylib next to the executable.
    * On Windows, this means copying the DLLs next to the executable. Sometimes users can't or won't change their PATH environment variable.

**Comment by tstenner** [comment]
1. I don't have any plans on how to handle dependencies, but the Apps I've been working on luckily need just Qt5, LSL and maybe Boost. `ExternalProject_Add` seems like a good idea, but we need to see where to add the code to do that (maybe single `.cmake` files in `/cmake`?)
2. By default, no apps are built, unless the corresponding `LSLAPPS_AppXY` is set. That way not even the build code for the unneeded apps is loaded. Auto opt-in if the dependencies are met would mean that we'd need to check the dependencies.
3. I totally agree :-) There's a lot of low hanging fruit (`Boost::thread`->`std::thread`, `boost::lexical_cast`->`std::to_string`) so some apps might be very easy convert if someone found the time.
For Linux / OS X with Homebrew a preinstalled boost should be easier, because it doesn't spew out pages of compiler warnings.
4. I'd prefer not to have every library in every app folder. Qt alone is about 10-20MB, so a typical install (Keyboard + LabRecorder + one app for a device) would already take up about 50MB.
   - OS X: I haven't tested it yet, but with relative rpaths the folders should be portable out of the box. With `../LSL/lib;lib/` users could copy the `LSL/lib` folder to the app path and have a portable app.
    - Windows: with the batch files, the change to the `PATH` is only temporary, but those who don't want that could copy the dlls to the app folder

**Comment by cboulay** [comment]
1. Yes, in `/cmake` sounds good. For each dependency, we can just define a macro that wraps Find<package>.cmake and if not found it uses `ExternalProject_Add`.
4. Yeah I guess it's fine to ask Windows users to use the .bat file. For my Mac-using collaborators, I'm usually the one that packages up the app so it's easy enough for me to also include a separate folder with the libs.

**Comment by dmedine** [comment]
The numbers here refer to the first post by cboulay and tstenner's response, not the post right above this one.

1. The reason for sticking the Qt libs in every folder is for packaging purposes. The idea is that researchers can just grab the two or three apps they need from the ftp and everything works right out of the box. I understand that this is a drag and causes bloat, but what you can do (and this is what I do) is only put the Qt libraries in the folders for the apps that you are developing. In fact, if you are not pushing anything to the ftp (and as far as I know, I am the only LSL developer that currently has access to the ftp server at the Swartz Center for Computational Neuroscience) you can simply put the Qt/lib directory on your PATH and Bob's your uncle. 

I break from boost's default file structure so that I can more easily choose between 32 and 64 bit version of the libraries. For me, what would be useful is to have a compile time option like boost_path=<path>. To me, this is better than an attempt at automation.

I'm happy to take the binaries out of each and every folder, but I don't want to give up the possibility to download them. The python script deps_worker.py can be used to download binaries for individual Apps or all of them. I don't want to lose this functionality. Also, I don't see any reason to ditch the python solution. It works quite well and it only requires small appending when a new app is created.

2. I think that XDF browser should also be built by default. I also think that we should revert back to the Python version of LabRecorder. Since cboulay fixed the pylsl bugs in osx, this makes sense because the scripting options would be available. Otherwise, whe should incorporate a lua interpreter into LabRecorder. This is low priority since not many people are asking for this kind of thing, but I thought I would mention it.

3. Replacing boost outright is actually not so easy. boost::thread can go, but a lot of the apps also use boost::algorithm::split. I'm not sure that has made its way into the STL (but maybeit did, I haven't looked very hard). Also, does the STL now support all of the overloads for lexical_cast? to_string is not the only one we need.  

Furthermore, while CMake is a nice build utility, VisualStudio has all the debugging tools that I need. I am happy to use valgrind and gdb on Linux, but for Windows, I need VisualStudio for its debugger. Thus, upgrading to C++14 is considerably more difficult than replacing a few lines of code. It requires updating vcproj files to vcxproj files and as far as I am aware, the only way to do this is to painstakingly rebuild all the projects settings in VisualStudio's (horrible) gui. 

Nevertheless, at some point this must be done. I have already started doing this on an app by app basis (I updated the Brain Products apps recently and performed this chore then). But, I would say that we should hold off on this and keep boost for now.

We should not replace boost's threading library with Qt's.

4. Only a handful of apps can be used on sytems other than Windows since the device drivers are only supported on Windows. I personally hate Apple and everything it stands for (greed, uncharitability, contempt for developers, misogyny, iPads, objective-C, the list goes on). So, I refuse touch it under any circumpstance. So please, as far as OSx or iOS goes, do what you will. 

But, maybe one idea for dealing with the rpath is [this](https://github.com/sccn/labstreaminglayer/blob/master/Apps/XDFBrowser/OSX/qtbuild.sh). Then you can cp the libs in the Contents/Resources folder and users can have joy just by double clicking the little apple package thing. Again, then you have this dylib bloat. But, at the end of the day, copying the lib into every app saves users trouble. If they need to have PATH set up just so, and for whatever reason it gets messed up, they are completely lost.

Users of LSL apps should never need to change the PATH, copy libraries around, or run .bat files. This is against the design specification (such as it is: as tstenner pointed out, we really need to clarify this now that more people are getting serious about development) of LSL. Users must be able simply to download a zip, open it up, click 'link' and be happy. We sacrifice elegance for ease of use here. Besides, any researcher that cares about having 5 copies of QtCore4.dll on her hardrive will have the know-how to organize her filesystem and set environment variables.

**Comment by tstenner** [comment]
1. I'll keep that in mind, in the worst case I can just create links on the build server and somehow let the zip program include the targets of the link so every archive contains the needed libraries. If we're going to distribute the libraries with the binaries anyway we could also use static linking for LSL and Qt on windows.
For Boost, it's a bit more complicated. With the current CMake file the path to Boost can be set with `-DBOOST_ROOT=c:/boost_xyz` but the included boost has to be handled separately (because it has to be built and doesn't have the CMake provided build targets).
2. I no one has started the Lua scripting yet, [QtScript](http://doc.qt.io/qt-5/qtscript-index.html) is integrated quite well with Qt and specified C++ classes.
3. I think a transition to header-only Boost libraries is a major win already, as Boost on Windows isn't as easy as it should be. C++11 should offer everything `lexical_cast` does with `std::to_string` for anything->string and `std::stof`, `std::stod` etc. for the other direction.

Instead of recreating the apps in Visual Studio, we should use the [CMake generators](https://cmake.org/cmake/help/v3.5/manual/cmake-generators.7.html#visual-studio-generators). Basically, you can run `cmake -G "Visual Studio 14 2015" path/to/LSL` and get a Visual Studio Project with a you can launch and use like a normal project.

**Comment by tstenner** [comment]
The code now builds aotmatically on [OS X](https://travis-ci.org/tstenner/labstreaminglayer/jobs/257599582), [Linux](https://travis-ci.org/tstenner/labstreaminglayer/jobs/257599581) and [Windows](https://ci.appveyor.com/project/tstenner/labstreaminglayer/build/1.0.64).

On Appveyor, we can download the resulting archive, on Travis CI we need either a server with FTP oder SCP access or put it on S3 (slightly more expensive). I tried `windeployqt` and so far it works ok, if we export a single archive with the whole tree from Travis the duplicated libraries should take no additional space in the archive but we'd need to extract it and generate several archives for distribution.

**Comment by cboulay** [comment]
Oh, and if we're worried about consistent SCCN FTP access, we can use GitHub's release pages, and make a Python script that automatically creates a release for us. [Example here](https://github.com/sccn/xdf/blob/master/make_release.py).

**Comment by tstenner** [comment]
I think I'll execute mac/windeployqt from within the python script, as CMake doesn't support target file names for install targets, only for build targets. It should also be necessary just once per directory, not per file but I haven't tested it yet with 2 Qt executables.

I'm currently compiling with Clang on Linux because (IIRC) the GCC version on Travis was slightly too old, it compiles faster on my machine and the diagnostics are more helpful. GCC also works, but I mostly use clang.

**Comment by cboulay** [comment]
I just gave it a go on Windows 10 with MSVC2015 x64.
For reference, here is my cmake command: `cmake -DBOOST_ROOT=C:\local\boost_1_61_0 -DBOOST_LIBRARYDIR=C:\local\boost_1_61_0\lib64-msvc-14.0 -DQt5_DIR=C:\Qt\5.7\msvc2015_64\lib\cmake\Qt5 -G "Visual Studio 14 2015 Win64" ..`.

Some thoughts...

* INSTALL.md talks about -DBOOST_ROOT but for most precompiled boost installs it will also be necessary to provide -DBOOST_LIBRARY_DIR.
    * I'm downloading boost 1.65 now and will try again.
* LSLAPPS_LabRecorder and LSLAPPS_XDFBrowser should default to ON (as per the discussion above). 
    * Otherwise there are warnings that the arguments are unused, as liblsl does not use them.
* Your instructions don't make it clear that `LSL_USE_SYSTEM_BOOST` is broken.
    * I'm downloading boost 1.65 now and will try again.
* The INSTALL target didn't work because it couldn't create the directory (error below). Maybe we just want to make the tree locally in labstreaminglayer/dist ?

```
7>  -- Installing: C:/Program Files/LabStreamingLayer/LSL/cmake
7>  CMake Error at cmake_install.cmake:31 (file):
7>    file INSTALL cannot make directory "C:/Program
7>    Files/LabStreamingLayer/LSL/cmake": No such file or directory
```

Overall it's very nice, but I still have a lot of testing to do. (Linux, MacOS, testing app portability, etc).

**Comment by cboulay** [comment]
On my linux box I actually tried using the `lslbuild.py` for the first time. I see that this provides a lot of arguments to the cmake command. I think the args/vals you chose are good and these options should be set by default even if not provided as arguments. This will make it easier for developers to use their preferred IDE without having to worry too much about these arguments. It'll then be trivially easy to point e.g. KDevelop or Qt Creator to the CMakeLists.txt file, or to use the Xcode generator, and set off on developing.

I ran into a linking error with LabRecorder on Linux. The solution is to put the zlib libraries in the link list AFTER boost. Here's the relevant change to Apps/LabRecorder/CMakeLists.txt:

```cmake
IF(MSVC)
    find_package(Boost REQUIRED COMPONENTS filesystem iostreams thread zlib)
    set(zlib_libs Boost::zlib)
ELSE()
    find_package(Boost REQUIRED COMPONENTS filesystem iostreams thread)
	find_package(ZLIB REQUIRED)
    set(zlib_libs ${ZLIB_LIBRARIES})
ENDIF()

target_link_libraries(${target}
	PRIVATE
	Qt5::Widgets
	Qt5::Network
	Boost::filesystem
	Boost::iostreams
	Boost::thread
    ${zlib_libs}
	LSL::lsl
)
```

**Comment by cboulay** [comment]
I was able to get windeployqt to work only on install following code I found [here](https://gitlab.pluribusgames.com/games/apitrace/blob/master/gui/CMakeLists.txt#L121-174). I've pushed my changes to your branch [here](https://github.com/cboulay/labstreaminglayer/tree/tstenner-cmake).

**Comment by tstenner** [comment]
Thanks for the extensive feedback, I already merged your changes.

I think instead of deploying the VC runtime DLLs with windeployqt we should include the redistributable installer (--no-system-runtime) manually and have new VC versions and especially other compilers on windows work out of the box (clang seems to be getting there, even with VS).

For default arguments we could put them in the root `CMakeLists.txt` (as in 8dd50c6568597f1cbb6d5b8c9311fc3ec6bf4094), but also for likely Boost / Qt-paths.

I still have some changes I need to merge back into the pull request and I hope to get to it on the weekend, some of them fix things with `LSL_USE_SYSTEM_BOOST`. In september I might be able to get my hands on a MacBook to test everything on it.

**Comment by cboulay** [comment]
I have a Macbook. I made some progress today (c7616110227dc1a94edc2b77b511c03720d37e42), but it's still not working perfectly. I'll try again tomorrow.

**Comment by cboulay** [comment]
I found the solution to the final problem [here](https://github.com/keepassxreboot/keepassxc/wiki/Set-up-Build-Environment-on-OS-X). It turns out that there is a problem with homebrew Qt.

The latest commit (c9244bf1068ec2d7de410dea1e98bc62d7594ee1) now seems to be working as good as I would like on all three major platforms. If you incorporate these changes into your PR then I'll be ready to merge whenever you are.

**Comment by tstenner** [comment]
That looks great. I'll try to wrap everything up on the weekend (+take a look at #212) and wait what David has to say about it.
Ideally, the CI builds will work by then and push to [Bintray](https://bintray.com/), but Bintray can wait.

**Comment by tstenner** [comment]
I rebased everything, so when everyone else is ready I'm ready.

**Comment by cboulay** [comment]
I made a PR to your branch to fix a couple bugs. With this latest commit, everything seems to be working well on Windows 10 / MSVC 2015 x64; MacOS Sierra Xcode (?); Ubunt 16.04 x64.

This PR shouldn't break anything @dmedine or @mgrivich do because they don't use cmake in their workflow. So as soon as you merge in my PR, then I can merge your PR into sccn.

Then I'll make sure pylsl and my other cmake-enabled apps can make use of the new structure.

**Comment by tstenner** [comment]
b38033997600ee8f530b956168e0e4c6f40b0abe removes some IDE projects, but apart from that the code mostly stays the same.

**Comment by dmedine** [comment]
This is both exciting and awesome! I tried to review the PR earlier 
today, but didn't have time to get very far. But, if cboulay is happy, 
then so am I. So my vote is to go ahead and merge. I seriously doubt 
anything substantial will need to be rolled back, but I'll hold on to 
the latest copy of the repository (prior to the merge) for a while so 
that if anything /does/ need to go back the way it was, we can pull in 
the necessary bits piece by piece.

And, let me personally extend a huge 'Thank You!' and initiate a huge 
round of applause for all the work that tstenner and cboulay put into 
this very, very necessary set of updates. Thanks, guys.

**Comment by cboulay** [comment]
@tstenner Thanks a lot for all your hard work on this.

Can you please provide some quick instructions on how we can setup CI builds on travis? I'm sure I can get all the details online, but I was hoping you already did most of the legwork and you just need to point me to the button.

**Comment by tstenner** [comment]
@cboulay It's surprisingly intuitive, I haven't tried setting a repository up as an organization.
Travis, AppVeyor (and in the future Bintray?) allow registrations with your Github-ID, that way the repositories and hopefully organizations etc. are set up automatically. For Travis and AppVeyor I included configurations (`.travis.yml` and `appveyor.yml?`) that get picked up automatically once you add a repository from your list to the CI. For Appveyor I haven't figured out how to build only some commits.
